### ISSUE TITLE
Inconsistent "TypeVar appears only once" & type solving behaviour between type alias & protocol

### ISSUE THREAD

**Issue by dchevell** [microsoft/pyright#11229]
**Describe the bug**
The full code example below will explain this better, but I'll try to summarise my use case:

I'm typing a decorator factory, where the outermost function takes some config parameters and returns an inner decorator. The inner decorator should only accept callables whose sole argument is a subtype of a specified class, so the inner function relies on a type variable that the outer function can't yet know about.

I've found that I can still use a type variable in the outermost function definition, even though it's not used in any parameters, because it's used twice in the return annotation: `Callable[[Callable[T], None]], Callable[[T], None]]` (see full example below). I can even assign that construct to a type variable, `Decorator[T]`. Pyright accepts this, and correctly solves the type variable and preserves the signature of the decorated function.

However, in my use case I need to use a more complex signature so I'm using a Protocol. When I define a protocol that's identical to the type alias, pyright gives me the error "TypeVar "T" appears only once", and my decorated function is solved as `(Unknown) -> None`

**Code or Screenshots**

Here's a self-contained example that demonstrates what I mean

```python
from collections.abc import Callable
from typing import Protocol, reveal_type


class A: ...

class B(A): ...


# Case 1

type Decorator1[T: A] = Callable[[Callable[[T], None]], Callable[[T], None]]

def deco_factory1[T: A](message: str) -> Decorator1[T]: # no warning
    def decorator(func: Callable[[T], None]) -> Callable[[T], None]:
        print(message)
        return func
    return decorator

@deco_factory1("test")
def test1(value: B) -> None: ...

reveal_type(test1) # information: Type of "test1" is "(B) -> None"


# Case 2

class Decorator2[T: A](Protocol):
    def __call__(self, func: Callable[[T], None], /) -> Callable[[T], None]: ...

def deco_factory2[T: A](message: str) -> Decorator2[T]: # warning: TypeVar "T" appears only once in generic function signature
    def deco(func: Callable[[T], None]) -> Callable[[T], None]:
        print(message)
        return func
    return deco

@deco_factory2("test")
def test2(value: B) -> None: ...

reveal_type(test2) # information: Type of "test2" is "(Unknown) -> None"
```

I would've expected the second case to give me the same results as the first. I can't think of any other workarounds for this sort of decorator factory pattern. I can't annotate my inner class as returning `Callable[[A], None]` because it's contravariant and wouldn't be able to wrap functions requiring a subtype of `A`, so I have to use a generic type (and therefore also can't annotate the outer function as `Callable[[Callable[[A], None], Callable[[A], None]]`). And I can't use `Callable[[Callable[..., None], Callable[..., None]]` on the outer function because it will erase the type of the decorated function.

I'm guessing this is a grey area of the typing spec, and wouldn't be surprised if Case 1 is technically unsupported as well, but the fact pyright solves case 1 correctly gives me a little hope.  Interestingly, mypy and ty can't solve the first case. Pyrefly can, but also fails on the second case.

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

Pyright (like mypy) contains some special casing for return types that are annotated with the `Callable` special form (or a type alias thereof). If a return `Callable` references a function-scoped type variable that is not referenced elsewhere in the function's signature, the TypeVar is rescoped to just that `Callable`. This same special case is not applied to callback protocols.

Admittedly, this special case is not part of the typing spec. However, there is enough existing code that depends on this behavior (which was originally introduced in mypy) that it would be painful to remove it now.
